### PR TITLE
design-proposal: cross-cluster mesh for tenant access to host services

### PR DESCRIPTION
**Pull request by kvaps**
## Summary

Adds a design proposal for cross-cluster connectivity between Cozystack-managed tenant clusters and the host cluster.

The motivating use case: a host cluster running Ceph (managed by Rook) that should be reachable from inside tenant clusters as if it were local storage. Standard single-gateway approaches (Submariner, Kilo's default `mesh-granularity=location`) bottleneck Ceph traffic; this proposal uses Kilo's `mesh-granularity=cross` (squat/kilo#328) to build a node-to-node mesh that scales linearly with cluster size and handles Rook-driven failover without controller intervention on the data path.

The proposal covers:
- Topology and why cross-mesh fits Ceph's traffic patterns
- A new operator (`cozystack-meshlink-operator`) and `TenantMeshLink` CRD for managing Peer objects on both sides
- Trust model (one-way: Cozystack → tenant; tenants have no host-cluster API access)
- IP allocation: no new address space; existing pod-CIDRs are sufficient
- Failure semantics, edge cases, and alternatives considered

Looking for feedback on the open questions, especially the upstream Kilo PR #328 strategy and whether tenant-side Kilo should be a hard requirement.

## Test plan

This is a design proposal; no code yet. Implementation testing is scoped in the proposal and will follow in implementation PRs:
- [ ] Unit tests for reconciliation logic
- [ ] Admission webhook tests for pod-CIDR overlap detection
- [ ] Integration tests with kind (two clusters)
- [ ] E2E with real Rook + tenant cluster + Ceph CSI

### CODE REVIEW

File: design-proposals/cross-cluster-tenant-mesh/README.md [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+
+1. The CRD does not contain references to tenants. Instead it accepts a map of peer clusters, each referenced by a Secret holding that cluster's kubeconfig. The Cozystack-specific notion of "tenant" lives entirely outside the controller.
+2. The Cozystack control plane integrates by translating its own tenant model into `ClusterMesh` objects (one per tenant connection, in the simple deployment shape) plus the corresponding kubeconfig Secrets. See [Cozystack integration](#cozystack-integration).
+
+## Goals
+
+- Pods in any peer cluster can reach selected services in another peer cluster as if they were on the local network. (Cozystack use case: tenant pods reach host Ceph monitors, OSDs and MDS daemons.)
+- Nodes added to or removed from any participating cluster are wired into / detached from the mesh automatically, without per-node manual configuration.
+- A compromise of a peer cluster (up to and including full root on a peer node) cannot affect routing in another peer cluster beyond the network surface that was explicitly granted, and cannot affect unrelated peers.
```
Comment (squat) [inline design-proposals/cross-cluster-tenant-mesh/README.md:58]: unless it is the cluster running the controller, in which case I guess they do get perms on peer clusters, but that's not new

File: design-proposals/cross-cluster-tenant-mesh/README.md [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+metadata:
+  name: example-mesh
+  namespace: kilo
+spec:
+  clusters:
+    cluster-a:
+      # The controller's own cluster — no kubeconfig needed.
+      local: true
+      allowedNetworks:
+        - 10.4.0.0/16          # WG-CIDR
```
Comment (squat) [inline design-proposals/cross-cluster-tenant-mesh/README.md:128]: Should any of these be named fields in the struct rather than open fields in `allowedNetworks`? If the WireGuard mesh CIDR and the Pod CIDR are mandatory then maybe they get special treatment? Alternatively, using the open list can later be easily migrated into the stricter design

Comment (squat) [reply design-proposals/cross-cluster-tenant-mesh/README.md:128]: I guess these are all technically optional and it just determines which networks from the Peer resources we want to honor / validate

File: design-proposals/cross-cluster-tenant-mesh/README.md [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+
+The CRD has no notion of tenants, hubs, spokes, owners, or trust direction. Asymmetric deployment shapes (e.g. one trusted cluster reachable by many less-trusted clusters) are expressed by the consumer of the controller through the choice of clusters they include in each `ClusterMesh` and through external RBAC on the kubeconfig Secrets.
+
+#### Reconciliation
+
+For each `ClusterMesh`, the controller performs validation in two layers — mesh-level checks that may halt the whole mesh, and per-Node checks that only skip the offending Node:
+
+**Mesh-level (halts reconciliation on failure):**
+
+1. Every CIDR in every `spec.clusters[*].allowedNetworks` is a subset of the controller's `--allowed-cidr` allowlist; otherwise `NetworksNotAllowed=True`.
```
Comment (squat) [inline design-proposals/cross-cluster-tenant-mesh/README.md]: This is kind of annoying, it means that there is functionally no difference between the cluster admin and the mesh admin. If you want to create a new mesh, then you have to edit the mesh controller deployment to add the allow list. I need to think about this a bit. What are we hoping to defend against here?

File: design-proposals/cross-cluster-tenant-mesh/README.md [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+
+```yaml
+apiVersion: kilo.squat.ai/v1alpha1
+kind: ClusterMesh
+metadata:
+  name: example-mesh
+  namespace: kilo
+spec:
+  clusters:
+    cluster-a:
```
Comment (squat) [inline design-proposals/cross-cluster-tenant-mesh/README.md]: Maybe in keeping with Kubernetes convention this should become a list of named structs, like how a Pod contains a list of named containers.

File: design-proposals/cross-cluster-tenant-mesh/README.md [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+- **Per-Node containment** validates that every observed annotation (`Node.Spec.PodCIDRs`, `kilo.squat.ai/wireguard-ip`) lies within the cluster's own `allowedNetworks`. A tenant root forging an annotation that points at the host pod-CIDR, host WG-CIDR, or any other CIDR the tenant did not declare itself is rejected — the offending Node is skipped and never appears as a Peer on the host side.
+- **Trust direction by kubeconfig placement.** Whichever cluster holds the controller and the kubeconfig Secrets is the side that drives writes; the side whose kubeconfig is held cannot write back. In Cozystack, only the host holds tenant kubeconfigs — trust flows host → tenant.
+- **Cross-mesh isolation.** Each `ClusterMesh`'s Peers are labelled with the mesh name; the controller never deletes or modifies Peers belonging to a different mesh, and `allowedNetworks` overlap between meshes (not just within a single mesh) is rejected.
```
Comment (squat) [inline design-proposals/cross-cluster-tenant-mesh/README.md]: We should probably also add labels for the source cluster name so if two controllers running on different hosts are managing meshes on the same tenant (some triangle where the two hosts don't know about each other) then the controllers are less likely to compete for ownership of Peers if the mesh object has the same name.

File: design-proposals/cross-cluster-tenant-mesh/README.md [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+2. **Cluster identifier scope**: should `spec.clusters` keys be free-form strings or follow a stricter schema (e.g. DNS-1123 labels) so they can be reused as label values? Likely the latter; to confirm during implementation.
+3. **Transitive routing**: with three or more clusters in the same `ClusterMesh`, the controller currently builds a full mesh. Should it support partial topologies (e.g. star)? Out of scope for v1; the CRD shape allows it later.
+4. **Multi-controller scenarios**: in a deployment where two clusters each run their own controller, how should they coordinate? Likely via a "leader" cluster identified in the CRD; deferred.
+5. **Per-peer opt-in for received CIDRs**: today `allowedNetworks` is a unilateral declaration on the source side, plus a global allowlist on the controller. Should there additionally be a per-peer `acceptedNetworks` field, so a peer can refuse to accept some of what another peer publishes? Likely unnecessary given the controller-level allowlist, but worth revisiting once there are multi-tenant deployments with heterogeneous policies.
```
Comment (squat) [inline design-proposals/cross-cluster-tenant-mesh/README.md]: The more I read about the controller allowlist, the more o actually started leaning in this direction. Maybe this needs to be a flag on Kilo, actually (or an entirely new PeerClass resource that declares what allowed IPs are permissible for every Peer in a cluster). This would allow individual clusters to guard against peers being created by a rogue cluster mesh controller. It's not blocking: this is orthogonal Kilo work that would be great to upstream to improve the administration of Kilo meshes.

**Review by kvaps — COMMENTED** [review]
_(withdrawn)_

**Comment by Arsolitt** [comment]
Two related scalability concerns that I think are missing from the proposal.

### Aggregate fan-out across tenants

The scaling analysis covers a single ClusterMesh, but on the host side peers from all tenants accumulate on every node:

`total_wg_peers_per_host_node = Σ(nodes_in_tenant_i)`

- **Small** (10 tenants × 10 nodes) → 100 peers/host node, ~3K Peer CRs in etcd
- **Medium** (50 × 15) → 750 peers/host node, ~22K Peer CRs
- **Large** (200 × 20) → 4,000 peers/host node, ~120K Peer CRs

WireGuard kernel datapath is fine (O(1) AllowedIPs lookup), but the control plane is a different story:

- **etcd**: tens of thousands of Peer objects, each generating watch events
- **kg-agent**: receives ALL Peers, rebuilds `wg setconf` on every change - at 4K peers that's ~200KB config reparsed per event
- **controller**: reconcile loop across all ClusterMesh CRs × all nodes × all peers

The proposal needs a scalability section with realistic limits and guidance on when to shard tenants across host clusters.

### Full bipartite without topology awareness

Every node in cluster A gets a tunnel to every node in cluster B. In heterogeneous clusters only a subset of nodes run the target workload - tunnels to irrelevant nodes carry zero useful traffic.

A `nodeSelector` on the cluster spec would scope the mesh to relevant nodes:

```yaml
spec:
  clusters:
    - name: host
      local: true
      allowedNetworks: [...]
      nodeSelector:
        storage-role: ceph
```
If 10 of 30 host nodes are selected, per-tenant peer count drops by 3x. Across 50 tenants: 750 → 250 peers per host node.

### Combined effect

These multiply: topology awareness reduces per-mesh fan-out, aggregate analysis tells you how many meshes you can sustain. Without both, there's no way to give operators capacity planning guidance.

File: design-proposals/cross-cluster-tenant-mesh/README.md (hunk 1 above)
Comment (kvaps) [reply design-proposals/cross-cluster-tenant-mesh/README.md:58]: Yes, intentional — that's the design. The controller cluster does have perms on remote Peers via the kubeconfigs it holds, and that's exactly the trust direction we want.

File: design-proposals/cross-cluster-tenant-mesh/README.md (hunk 2 above)
Comment (kvaps) [reply design-proposals/cross-cluster-tenant-mesh/README.md:128]: Kept the flat list for v1alpha1 — it's intentionally narrow ('nothing else than what's listed can end up in a Peer') and easy to migrate to typed fields later if needed. Recorded as Open Question #6 in 8830a7c.

Comment (kvaps) [reply design-proposals/cross-cluster-tenant-mesh/README.md:128]: Right, exactly. The list is purely about what's permissible / honoured — not categorising CIDRs into kinds. That's why we kept it flat.

File: design-proposals/cross-cluster-tenant-mesh/README.md (hunk 3 above)
Comment (kvaps) [reply design-proposals/cross-cluster-tenant-mesh/README.md]: Agreed, and dropped. The `--allowed-cidr` allowlist is gone in 8830a7c. The current design relies on RBAC on `ClusterMesh` as the address-surface chokepoint, and cluster-side defense-in-depth is now tracked in Alternatives as future Kilo work (your `PeerClass` suggestion below).

File: design-proposals/cross-cluster-tenant-mesh/README.md (hunk 4 above)
Comment (kvaps) [reply design-proposals/cross-cluster-tenant-mesh/README.md]: Done in 8830a7c — `spec.clusters` is now a list of named entries (`- name: cluster-a, ...`).

File: design-proposals/cross-cluster-tenant-mesh/README.md (hunk 5 above)
Comment (kvaps) [reply design-proposals/cross-cluster-tenant-mesh/README.md]: Added in 8830a7c — every generated Peer now carries `kilo-clustermesh.io/mesh: <mesh-name>` and `kilo-clustermesh.io/source-cluster: <cluster-name>`, with the controller's ownership scoped to its own (mesh, source-cluster) pairs. Two controllers can coexist on the same remote cluster as long as they're not declaring the same source cluster.

File: design-proposals/cross-cluster-tenant-mesh/README.md (hunk 6 above)
Comment (kvaps) [reply design-proposals/cross-cluster-tenant-mesh/README.md]: Agreed — tracked in Alternatives as future Kilo work in 8830a7c, with credit to you. Independent of this `ClusterMesh` proposal but a clear next direction for hardening Kilo administration.
